solver/src/analytics.rs

```diff
@@ -0,0 +1,126 @@
+        .map(|trade| {
+            let sell_token_price = &submitted_prices[&trade.order.order_creation.sell_token];
+            let buy_token_price = &submitted_prices[&trade.order.order_creation.buy_token];
+            (
+                trade.order.order_meta_data.uid,
+                SurplusInfo {
+                    solver_name: winning_solver,
+                    absolute: trade
+                        .surplus(sell_token_price, buy_token_price)
+                        .unwrap_or_else(BigRational::zero),
+                    ratio: trade
+                        .surplus_ratio(sell_token_price, buy_token_price)
+                        .unwrap_or_else(BigRational::zero),
+                },
+            )
+        })
+        .collect();
+
+    let best_alternative = best_surplus_by_order(alternative_settlements);
+
+    for (order_id, submitted) in submitted_surplus.iter() {
+        if let Some(alternative) = best_alternative.get(order_id) {
+            metrics.report_order_surplus(
+                winning_solver,
+                alternative.solver_name,
+                (&submitted.ratio - &alternative.ratio)
+                    .to_f64()
+                    .unwrap_or_default(),
+            );
+            if alternative.ratio > submitted.ratio {
+                tracing::warn!(
+                    ?order_id, %submitted, %alternative,
+                    "submission surplus worse than lower ranked settlement",
+                );
+            }
+        }
+    }
+}
+
+fn best_surplus_by_order(
+    settlements: Vec<(&'static str, &Settlement)>,
+) -> HashMap<OrderUid, SurplusInfo> {
+    let mut best_surplus: HashMap<OrderUid, SurplusInfo> = HashMap::new();
+    for (solver, settlement) in settlements.iter() {
+        let trades = settlement.trades();
+        let clearing_prices = get_prices(settlement);
+        for trade in trades {
+            let order_id = trade.order.order_meta_data.uid;
+            let sell_token_price = &clearing_prices[&trade.order.order_creation.sell_token];
+            let buy_token_price = &clearing_prices[&trade.order.order_creation.buy_token];
+            let surplus = SurplusInfo {
+                solver_name: solver,
+                absolute: trade
+                    .surplus(sell_token_price, buy_token_price)
+                    .unwrap_or_else(BigRational::zero),
+                ratio: trade
+                    .surplus_ratio(sell_token_price, buy_token_price)
+                    .unwrap_or_else(BigRational::zero),
+            };
+            let entry = best_surplus.entry(order_id);
+            match entry {
+                Entry::Occupied(mut entry) => {
+                    let value = entry.get_mut();
+                    if value.absolute < surplus.absolute {
+                        *value = surplus;
+                    }
+                }
+                Entry::Vacant(entry) => {
+                    entry.insert(surplus);
+                }
+            }
```

Comment:
ubernit: Note that I think this can be written more succinctly with:
```suggestion
            best_surplus.entry(order_id)
                .and_modify(|entry| {
                    if entry.absolute < surplus.absolute {
                        *entry = surplus.clone();
                    }
                })
                .or_insert(surplus);
```